src/mongo/db/pipeline/expression.cpp

```diff
@@ -3072,6 +3085,150 @@ const char* ExpressionWeek::getOpName() const {
     return "$week";
 }
 
+/* ------------------------- ExpressionIsoDayOfWeek --------------------- */
+
+Value ExpressionIsoDayOfWeek::evaluateInternal(Variables* vars) const {
+    Value date(vpOperand[0]->evaluateInternal(vars));
+    return Value(extract(date.coerceToTm()));
+}
+
+int ExpressionIsoDayOfWeek::extract(const tm& tm) {
+    // translate from sunday=0 … saturday=6 to monday=1 … sunday=7
+    return (tm.tm_wday - 7) % 7 + 7;
+}
+
+REGISTER_EXPRESSION(isoDayOfWeek, ExpressionIsoDayOfWeek::parse);
+const char* ExpressionIsoDayOfWeek::getOpName() const {
+    return "$isoDayOfWeek";
+}
+
+/* ------------------------- ExpressionIsoWeekYear ---------------------- */
+
+Value ExpressionIsoWeekYear::evaluateInternal(Variables* vars) const {
+    Value date(vpOperand[0]->evaluateInternal(vars));
+    return Value(extract(date.coerceToTm()));
+}
+
+int ExpressionIsoWeekYear::extract(const tm& tm) {
+    if (tm.tm_mon > 0 && tm.tm_mon < 11) {
+        // If month is between February and November, it is just the year given.
+        return tm.tm_year + 1900;
+    } else if (tm.tm_mon == 0) {
+        // In January we need to check if the week belongs to previous year.
+        int isoWeek = ExpressionIsoWeek::extract(tm);
+        if (isoWeek > 51) { // Weeks 52 and 53 belong to the previous year.
+            return tm.tm_year + 1900 - 1;
+        } else { // All other weeks belong to given year.
+            return tm.tm_year + 1900;
+        }
+    } else {
+        // A week 1 in December belongs to the next year.
+        int isoWeek = ExpressionIsoWeek::extract(tm);
+        if (isoWeek == 1) {
+            return tm.tm_year + 1900 + 1;
+        } else {
+            return tm.tm_year + 1900;
+        }
+    }
+}
+
+REGISTER_EXPRESSION(isoWeekYear, ExpressionIsoWeekYear::parse);
+const char* ExpressionIsoWeekYear::getOpName() const {
+    return "$isoWeekYear";
+}
+
+/* ------------------------- ExpressionIsoWeek -------------------------- */
+
+Value ExpressionIsoWeek::evaluateInternal(Variables* vars) const {
+    Value date(vpOperand[0]->evaluateInternal(vars));
+    return Value(extract(date.coerceToTm()));
+}
+
+int ExpressionIsoWeek::lastWeek(int year) {
+    // Create YYYY-12-31T23:59:59 so only 1 second left to new year.
+    struct tm tm;
+    tm.tm_year = year - 1900;
+    tm.tm_mon = 11;
+    tm.tm_mday = 31;
+    tm.tm_hour = 23;
+    tm.tm_min = 59;
+    tm.tm_sec = 59;
+    tm.tm_wday = -1; // this was enough to solve the bug for 1900
+    tm.tm_yday = -1;
+    tm.tm_gmtoff = 0;
+    tm.tm_isdst = 0;
```

Comment:
Unfortunately, this doesn't work on windows, which is one of our supported platforms.
```
Compiling build\cached\mongo\db\pipeline\expression.obj
expression.cpp
src\mongo\db\pipeline\expression.cpp(3127) : error C2039: 'tm_gmtoff' : is not a member of 'tm'
        C:\Program Files (x86)\Microsoft Visual Studio 12.0\VC\INCLUDE\wchar.h(1170) : see declaration of 'tm'
scons: *** [build\cached\mongo\db\pipeline\expression.obj] Error 2
scons: building terminated because of errors.
```
Here's the definition:
```c
#ifndef _TM_DEFINED
struct tm {
 int tm_sec; /* seconds after the minute - [0,59] */
 int tm_min; /* minutes after the hour - [0,59] */
 int tm_hour; /* hours since midnight - [0,23] */
 int tm_mday; /* day of the month - [1,31] */
 int tm_mon; /* months since January - [0,11] */
 int tm_year; /* years since 1900 */
 int tm_wday; /* days since Sunday - [0,6] */
 int tm_yday; /* days since January 1 - [0,365] */
 int tm_isdst; /* daylight savings time flag */
 };
#define _TM_DEFINED
#endif /* _TM_DEFINED */
```

This was fixed for me by changing the initialization of tm: 
```c++
struct tm tm = {};  // This will initialize all struct members with 0.
```
Then removing lines 3158 and 3159. See [this question on stackoverflow](http://stackoverflow.com/questions/1069621/are-members-of-a-c-struct-initialized-to-0-by-default) for struct initialization details.